lib/src/utils.dart

```diff
@@ -1,7 +1,147 @@
 // Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
+import 'dart:math' as math;
 
 /// Pads [source] to [length] by adding spaces at the end.
 String padRight(String source, int length) =>
     source + ' ' * (length - source.length);
+
+/// Wraps a block of text into lines no longer than [length].
+///
+/// Tries to split at whitespace, but if that's not good enough to keep it
+/// under the limit, then it splits in the middle of a word.
+///
+/// Preserves indentation (leading whitespace) for each line (delimited by '\n')
+/// in the input, and will indent wrapped lines the same amount.
+///
+/// If [hangingIndent] is supplied, then that many spaces will be added to each
+/// line, except for the first line. This is useful for flowing text with a
+/// heading prefix (e.g. "Usage: "):
+///
+/// ```dart
+/// var prefix = "Usage: ";
+/// print(prefix + wrapText(invocation, hangingIndent: prefix.length, length: 40));
+/// ```
+///
+/// yields:
+/// ```
+/// Usage: app main_command <subcommand>
+///        [arguments]
+/// ```
+///
+/// If [length] is not specified, then no wrapping occurs, and the original
+/// [text] is returned unchanged.
+String wrapText(String text, {int length, int hangingIndent}) {
+  if (length == null) return text;
+  hangingIndent ??= 0;
+  var splitText = text.split('\n');
+  var result = <String>[];
+  for (var line in splitText) {
+    var trimmedText = line.trimLeft();
+    final leadingWhitespace =
+        line.substring(0, line.length - trimmedText.length);
+    var notIndented;
+    if (hangingIndent != 0) {
+      // When we have a hanging indent, we want to wrap the first line at one
+      // width, and the rest at another (offset by hangingIndent), so we wrap
+      // them twice and recombine.
+      final firstLineWrap = wrapTextAsLines(
+        trimmedText,
+        length: length - leadingWhitespace.length,
+      );
+      notIndented = <String>[firstLineWrap.removeAt(0)];
+      trimmedText = trimmedText.substring(notIndented[0].length).trimLeft();
+      if (firstLineWrap.isNotEmpty) {
+        notIndented.addAll(wrapTextAsLines(
+          trimmedText,
+          length: length - leadingWhitespace.length - hangingIndent,
+        ));
+      }
+    } else {
+      notIndented = wrapTextAsLines(
+        trimmedText,
+        length: length - leadingWhitespace.length,
+      );
+    }
+    String hangingIndentString;
+    result.addAll(notIndented.map<String>(
+      (String line) {
```

Comment:
The type annotation isn't needed.

https://www.dartlang.org/guides/language/effective-dart/design#avoid-annotating-inferred-parameter-types-on-function-expressions

Also, can you dartfmt this?